modules/highgui/src/window_wayland.cpp

```diff
@@ -0,0 +1,2482 @@
+    wl_display_flush(display_);
+    input_.reset();
+    wl_display_disconnect(display_);
+}
+
+void cv_wl_display::init(const char *display) {
+    display_ = wl_display_connect(display);
+    if (!display_)
+        throw_system_error("Could not connect to display", errno)
+
+    registry_ = wl_display_get_registry(display_);
+    wl_registry_add_listener(registry_, &reg_listener_, this);
+    wl_display_roundtrip(display_);
+    if (!compositor_ || !shm_ || !xdg_wm_base_ || !input_)
+        CV_Error(StsInternal, "Compositor doesn't have required interfaces");
+
+    wl_display_roundtrip(display_);
+    if (!(formats_ & (1 << WL_SHM_FORMAT_XRGB8888)))
+        CV_Error(StsInternal, "WL_SHM_FORMAT_XRGB32 not available");
+
+    poller_.add(
+            wl_display_get_fd(display_),
+            EPOLLIN | EPOLLOUT | EPOLLERR | EPOLLHUP
+    );
+}
+
+int cv_wl_display::dispatch() {
+    return wl_display_dispatch(display_);
+}
+
+int cv_wl_display::dispatch_pending() {
+    return wl_display_dispatch_pending(display_);
+}
+
+int cv_wl_display::flush() {
+    return wl_display_flush(display_);
+}
+
+int cv_wl_display::roundtrip() {
+    return wl_display_roundtrip(display_);
+}
+
+std::pair<uint32_t, bool> cv_wl_display::run_once(int timeout) {
+    // prepare to read events
+    this->dispatch_pending();
+    int ret = this->flush();
+    if (ret < 0 && errno == EAGAIN) {
+        poller_.modify(wl_display_get_fd(display_),
+                       EPOLLIN | EPOLLOUT | EPOLLERR | EPOLLHUP);
+    } else if (ret < 0) {
+        return std::make_pair(0, false);
+    }
+
+    auto events = poller_.wait(timeout);
+    if (events.empty())
+        return std::make_pair(0, true);
+
+    uint32_t events_ = events[0].events;
+    if (events_ & EPOLLIN) {
+        this->dispatch();
+    }
+
+    if (events_ & EPOLLOUT) {
+        if (this->flush() == 0) {
+            poller_.modify(wl_display_get_fd(display_),
+                           EPOLLIN | EPOLLERR | EPOLLHUP);
+        }
+    }
+    return std::make_pair(events_, false);
+}
+
+struct wl_shm *cv_wl_display::shm() {
+    return shm_;
+}
+
+weak_ptr<cv_wl_input> cv_wl_display::input() {
+    return input_;
+}
+
+uint32_t cv_wl_display::formats() const {
+    return formats_;
+}
+
+struct wl_surface *cv_wl_display::get_surface() {
+    return wl_compositor_create_surface(compositor_);
+}
+
+struct xdg_surface *cv_wl_display::get_shell_surface(struct wl_surface *surface) {
+    return xdg_wm_base_get_xdg_surface(xdg_wm_base_, surface);
+}
+
+void cv_wl_display::handle_reg_global(void *data, struct wl_registry *reg, uint32_t name, const char *iface,
+                                      uint32_t version) {
+    std::string const interface = iface;
+    auto *display = reinterpret_cast<cv_wl_display *>(data);
+
+    if (interface == wl_compositor_interface.name) {
+        display->compositor_ = (struct wl_compositor *)
+                wl_registry_bind(reg, name, &wl_compositor_interface, version);
```

Comment:
need to do the min-version dance for all interfaces - compositor, SHM, seat - if you tell the server you support higher versions than the client is explicitly aware of, then it'll break